### PR TITLE
Handle onion request/response packets

### PR DESCRIPTION
**Pull request by kurnevsky**

### CODE REVIEW

File: src/toxcore/dht/server.rs [hunk 1]
```diff
@@ -435,6 +463,159 @@ impl Server {
...
+    /**
+    handle received OnionRequest2 packet, then create AnnounceRequest
+    or OnionDataRequest packet and send it to the next peer.
+    */
+    pub fn handle_onion_request_2(&self, packet: OnionRequest2, addr: SocketAddr) -> IoFuture<()> {
+        let onion_symmetric_key = self.onion_symmetric_key.read();
+        let shared_secret = precompute(&packet.temporary_pk, &self.sk);
+        if let Ok(payload) = packet.get_payload(&shared_secret) {
+            let onion_return = OnionReturn::new(
+                &onion_symmetric_key,
```
[the remark this answers is not in the document]
Comment (kurnevsky) [reply src/toxcore/dht/server.rs]: All other packets (including dht packets) take keys by reference. So I'd suggest not to touch it in this PR and if we decide to use vaulues change it everywhere.

File: src/toxcore/dht/server.rs [hunk 2]
```diff
@@ -872,6 +1057,410 @@ mod tests {
...
+        };
+        let packet = DhtPacket::OnionRequest0(OnionRequest0::new(&precomp, &bob_pk, payload));
+
+        assert!(alice.handle_packet((packet, addr)).wait().is_ok());
+
+        let (packet, _rx) = rx.into_future().wait().unwrap();
+
+        let (next, addr_to_send) = match packet {
+            Some((DhtPacket::OnionRequest1(next), addr_to_send)) => (next, addr_to_send),
+            p => panic!("Expected OnionRequest1 but got {:?}", p)
```
Comment (kpp) [inline src/toxcore/dht/server.rs]: You will get a lot of missing covered lines on these `p => panic!`

**Comment by kpp** [comment]
So far so good

File: src/toxcore/dht/server.rs [hunk 3]
```diff
@@ -872,6 +1057,410 @@ mod tests {
...
+    }
+
+    #[test]
+    fn server_handle_onion_request_0_invalid_payload_test() {
+        let (alice, _precomp, _bob_pk, _bob_sk, _rx, addr) = create_node();
+
+        let packet = DhtPacket::OnionRequest0(OnionRequest0 {
+            nonce: gen_nonce(),
+            temporary_pk: gen_keypair().0,
+            payload: vec![42, 123]
```
[the remark this answers is not in the document]
Comment (kurnevsky) [reply src/toxcore/dht/server.rs]: Because it's not encrypted with `onion_symmetric_key`.

[the remark this answers is not in the document]
Comment (kpp) [reply src/toxcore/dht/server.rs]: please comment it.

**Pull request merged**